src/ring_theory/witt_vector/is_poly.lean

```diff
@@ -0,0 +1,627 @@
+is said to be *polynomial* if there is a family of polynomials `φₙ` over `ℤ` such that the `n`th
+coefficient of `f x y` is given by evaluating `φₙ` at the coefficients of `x` and `y`.
+
+See also `witt_vector.is_poly` for the unary variant.
+
+The `ghost_calc` tactic treats `is_poly₂` as a type class,
+and the `@[is_poly]` attribute derives certain specialized composition instances
+for declarations of type `is_poly₂ f`.
+For the most part, users are not expected to treat `is_poly₂` as a class.
+-/
+@[class] def is_poly₂ (f : Π ⦃R⦄ [comm_ring R], witt_vector p R → 𝕎 R → 𝕎 R) : Prop :=
+∃ φ : ℕ → mv_polynomial (fin 2 × ℕ) ℤ, ∀ ⦃R⦄ [comm_ring R] (x y : 𝕎 R),
+  by exactI (f x y).coeff = λ n, peval (φ n) ![x.coeff, y.coeff]
+
+
+variable {p}
+
+/-- The composition of polynomial functions is polynomial. -/
+lemma is_poly₂.comp {h f g} (hh : is_poly₂ p h) (hf : is_poly p f) (hg : is_poly p g) :
+  is_poly₂ p (λ R _Rcr x y, by exactI h (f x) (g y)) :=
+begin
+  unfreezingI
+  { obtain ⟨φ, hf⟩ := hf,
+    obtain ⟨ψ, hg⟩ := hg,
+    obtain ⟨χ, hh⟩ := hh },
+  refine ⟨(λ n, bind₁ (uncurry $
+          ![λ k, rename (prod.mk (0 : fin 2)) (φ k),
+            λ k, rename (prod.mk (1 : fin 2)) (ψ k)]) (χ n)), _⟩,
+  intros,
+  funext n,
+  simp only [peval, aeval_bind₁, function.comp, hh, hf, hg, uncurry],
+  apply eval₂_hom_congr rfl _ rfl,
+  ext ⟨i, n⟩,
+  fin_cases i;
+  simp only [aeval_eq_eval₂_hom, eval₂_hom_rename, function.comp, matrix.cons_val_zero,
+    matrix.head_cons, matrix.cons_val_one],
+end
+
+/-- The composition of a polynomial function with a binary polynomial function is polynomial. -/
+lemma is_poly.comp₂ {g f} (hg : is_poly p g) (hf : is_poly₂ p f) :
+  is_poly₂ p (λ R _Rcr x y, by exactI g (f x y)) :=
+begin
+  unfreezingI
+  { obtain ⟨φ, hf⟩ := hf,
+    obtain ⟨ψ, hg⟩ := hg },
+  use (λ n, bind₁ φ (ψ n)),
+  intros,
+  simp only [peval, aeval_bind₁, function.comp, hg, hf]
+end
+
+/-- The diagonal `λ x, f x x` of a polynomial function `f` is polynomial. -/
+lemma is_poly₂.diag {f} (hf : is_poly₂ p f) :
+  is_poly p (λ R _Rcr x, by exactI f x x) :=
+begin
+  unfreezingI {obtain ⟨φ, hf⟩ := hf},
+  refine ⟨λ n, bind₁ (uncurry ![X, X]) (φ n), _⟩,
+  intros, funext n,
+  simp only [hf, peval, uncurry, aeval_bind₁],
+  apply eval₂_hom_congr rfl _ rfl,
+  ext ⟨i, k⟩, fin_cases i;
+  simp only [matrix.head_cons, aeval_X, matrix.cons_val_zero, matrix.cons_val_one],
+end
+
+namespace tactic
+open tactic
+
+/-!
+### The `@[is_poly]` attribute
+
+This attribute is used to derive specialized composition instances
+for `is_poly` and `is_poly₂` declarations.
+-/
+
+/--
+If `n` is the name of a lemma with opened type `∀ vars, tp`,
+where `tp` is an application of `is_poly`,
+`mk_poly_comp_lemmas n vars tp` adds composition instances to the environment
+`n.comp_i` and `n.comp₂_i`.
+-/
+meta def mk_poly_comp_lemmas (n : name) (vars : list expr) (tp : expr) : tactic unit :=
+do c ← mk_const n,
+   `(is_poly %%p _) ← return tp,
```

Comment:
This is an idiom I have not seen before, is it the same as this:
```suggestion
meta def mk_poly_comp_lemmas (n : name) (vars : list expr) : tactic unit
| `(is_poly %%p _) :=
do c ← mk_const n,
```
(with a catch-all case `| _ := failure` presumably?)

Reply:
Yes, in general the `pattern <- t` syntax is short for 
```lean
t' <- t, 
match t' with 
| pattern := ...
| _ := failed
end
```

With the change you suggested below, this is irrelevant. Thanks!